### PR TITLE
Implementing Reflect() boundaries

### PR DESCRIPTION
**Pull request by nicolasmerino41**
Hey, there,

I've already made some changes. I've noticed that for wrap() update_boundaries in array.jl, this code:
```
function update_boundary!(A::AbstractStencilArray{S,R}, ::Halo, ::Wrap) where {S<:Tuple{L},R} where {L}
    src = parent(A)
    startpad = 1:R
    endpad = L+R+1:L+2R
    startvals = R+1:2R
    endvals = L+1:L+R
    @assert length(startpad) == length(endvals) == R
    @assert length(endpad) == length(startvals) == R
    @inbounds src[startpad] .= src[endvals]
    @inbounds src[endpad] .= src[startvals]
    return A
end
```
And this code: 
```
function update_boundary!(A::AbstractStencilArray{S,R}, ::Halo, ::Wrap) where {S<:Tuple{Y,X},R} where {Y,X}
    src = parent(A)
    n_xs, n_ys = X, Y
    startpad_x = startpad_y = 1:R
    endpad_x = n_xs+R+1:n_xs+2R
    endpad_y = n_ys+R+1:n_ys+2R
    start_x = start_y = R+1:2R
    end_x = n_xs+1:n_xs+R
    end_y = n_ys+1:n_ys+R
    xs = 1:n_xs+2R
    ys = 1:n_ys+2R

    @assert length(startpad_x) == length(start_x) == R
    @assert length(endpad_x) == length(end_x) == R
    @assert length(startpad_y) == length(start_y) == R
    @assert length(endpad_y) == length(end_y) == R
    @assert map(length, (ys, xs)) === size(src)

    CI = CartesianIndices
    # Sides ---
    @inbounds src[CI((ys, startpad_x))] .= src[CI((ys, end_x))]
    @inbounds src[CI((ys, endpad_x))]   .= src[CI((ys, start_x))]
    @inbounds src[CI((startpad_y, xs))] .= src[CI((end_y, xs))]
    @inbounds src[CI((endpad_y, xs))]   .= src[CI((start_y, xs))]

    # Corners ---
    @inbounds src[CI((startpad_y, startpad_x))] .= src[CI((end_y, end_x))]
    @inbounds src[CI((startpad_y, endpad_x))]   .= src[CI((end_y, start_x))]
    @inbounds src[CI((endpad_y, startpad_x))]   .= src[CI((start_y, end_x))]
    @inbounds src[CI((endpad_y, endpad_x))]     .= src[CI((start_y, start_x))]

    return after_update_boundary!(A)
end
```
Seem to be doing the same? I don't fully understand it. So for Reflect() I only followed the second approach.

### CODE REVIEW

**Comment by nicolasmerino41** [comment]
Nevermind about the two approaches. I understand they're for 1D and 2D, respectively.

File: Manifest.toml [hunk 1]
```diff
@@ -0,0 +1,300 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (rafaqz) [inline Manifest.toml:1]: You will need to remove this file

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (rafaqz) [inline .vscode/settings.json:1]: Remove this

File: src/array.jl [hunk 3]
```diff
@@ -261,12 +261,111 @@ function update_boundary!(A::AbstractStencilArray{S,R}, ::Halo, ::Wrap) where {S
     @inbounds src[CI((startpad_y, endpad_x, startpad_z))] .= src[CI((end_y, start_x, end_z))]
     @inbounds src[CI((startpad_y, endpad_x, endpad_x))] .= src[CI((end_y, start_x, start_z))]
     @inbounds src[CI((endpad_y, endpad_x, endpad_z))] .= src[CI((start_y, start_x, start_z))]
-    @inbounds src[CI((endpad_y, startpad_x, endpad_z))] .= src[CI((end_y, start_x, start_z))]
+    @inbounds src[CI((endpad_y, startpad_x, endpad_z))] .= src[CI((start_y, end_x, start_z))]
```
Comment (rafaqz) [inline src/array.jl:264]: Why switch this?

Comment (nicolasmerino41) [reply src/array.jl:264]: I think it should go the way I wrote it. Am I wrong? It didn't follow the pattern you were using for the other corners

Comment (rafaqz) [reply src/array.jl:264]: Ok no worries. Normally we try not to change unrelated things in PRs to keep the intent clear, maybe leave a TODO comment instead so I can look at it later

Comment (nicolasmerino41) [reply src/array.jl:264]: Make sense. Sorry bout that :)

Comment (rafaqz) [reply src/array.jl:264]: No worries, just housekeeping. The PR is great

**Comment by nicolasmerino41** [comment]
Ok, so I added the export Reflect. Also implemented getneighbor(::Reflect). I was tempted to create a custom unsafe_getneighbor specific for Reflect because I wasn't sure it'd behave properly with the default one:
```
@inline function unsafe_getneighbor(
    A::AbstractStencilArray{<:Any,R}, ::BoundaryCondition, pad::Padding, I::Tuple
) where R
    unsafe_getindex(A, pad, I...)
end
```
But gave up on it.
Regardless of the additions, the test still doesn't pass, and now I just simplified it to:
@testset "trial" begin
    r = (1.0:5.0) * (100.0:105.0)'
    @test A = StencilArray(r, Window{1,2}(); padding=Conditional(), boundary=Reflect())
end
Which is exclusively trying to build the StencilArray, so clearly it's something about the definition of Reflect() somewhere, but I have no idea where that should be. I've checked everywhere Remove and Wrap are defined so...

File: src/array.jl [hunk 4]
```diff
@@ -133,6 +133,14 @@ end
 @inline function getneighbor(A::AbstractStencilArray, boundary::Remove, ::Conditional, I::Tuple)
     checkbounds(Bool, A, I...) ? (@inbounds A[I...]) : boundary.padval
 end
+# For Reflect we mirror the index around the boundary
+@inline function getneighbor(A::AbstractStencilArray{S,R}, ::Reflect, pad::Conditional, I::Tuple) where {S,R}
+    sz = tuple_contents(S)
+    reflected_inds = map(I, sz) do i, s
+        i < 1 ? 2 - i : (i > s ? 2*s - i : i)
```
Comment (rafaqz) [inline src/array.jl:140]: Can we use if else here instead? The double conditional on one line is hard to follow

Comment (nicolasmerino41) [reply src/array.jl]: Of course, here you have it. Given that I was actually following the syntax used for Wrap(), I took the freedom of rewriting it for Wrap too, hope that's ok :)

Comment (rafaqz) [reply src/array.jl]: Haha thanks.

**Comment by rafaqz** [comment]
Thanks! I'll bump a patch version so we can get the other PR in
